### PR TITLE
Add support for deep=True to `cirq.eject_phased_paulis` transformer

### PR DESCRIPTION
**Pull request by tanujkhattar**
- Adds support to recursively run `cirq.eject_phased_paulis` transformer on circuits wrapped inside a circuit operation by setting deep=True in transformer context.
- Part of https://github.com/quantumlib/Cirq/issues/5039

### CODE REVIEW

File: cirq-core/cirq/transformers/eject_phased_paulis.py [hunk 1]
```diff
@@ -66,6 +66,7 @@ def map_func(op: 'cirq.Operation', _: int) -> 'cirq.OP_TREE':
 
         # Collect, phase, and merge Ws.
         w = _try_get_known_phased_pauli(op, no_symbolic=not eject_parameterized)
+        # print(f"DEBUG1 {_}: ", op, w, sep="\n")
```
Comment (MichaelBroughton) [inline cirq-core/cirq/transformers/eject_phased_paulis.py:69]: Remove.

File: cirq-core/cirq/transformers/eject_phased_paulis.py [hunk 2]
```diff
@@ -74,6 +75,7 @@ def map_func(op: 'cirq.Operation', _: int) -> 'cirq.OP_TREE':
             )
 
         affected = [q for q in op.qubits if q in held_w_phases]
+        # print(f"DEBUG2 {_}: ", op, affected, sep="\n")
```
Comment (MichaelBroughton) [inline cirq-core/cirq/transformers/eject_phased_paulis.py:78]: remove.

File: cirq-core/cirq/transformers/eject_phased_paulis_test.py [hunk 3]
```diff
@@ -53,29 +54,42 @@ def assert_optimizes(
...
-        "{!r}"
-    ).format(before, expected, circuit, expected, circuit)
+    cirq.testing.assert_same_circuits(circuit, expected)
 
     # And it should be idempotent.
     circuit = cirq.eject_phased_paulis(
         circuit, eject_parameterized=eject_parameterized, context=context
     )
-    assert circuit == expected
+    cirq.testing.assert_same_circuits(circuit, expected)
```
Comment (MichaelBroughton) [inline cirq-core/cirq/transformers/eject_phased_paulis_test.py:63]: Move here and below to own test ?

**Review by MichaelBroughton — APPROVED** [review]
Few tweaks then LGTM

File: cirq-core/cirq/transformers/eject_phased_paulis_test.py (hunk 3 above)
Comment (tanujkhattar) [reply cirq-core/cirq/transformers/eject_phased_paulis_test.py:63]: The nested cases are pretty complex (as we have already seen in the bugfixes in previous PRs), so I prefer this approach where we test the nested case for every non-nested case (eg: with / without ignore tags, with / without eject_parameterized etc.)

File: cirq-core/cirq/transformers/eject_phased_paulis.py (hunk 1 above)
Comment (tanujkhattar) [reply cirq-core/cirq/transformers/eject_phased_paulis.py]: done

File: cirq-core/cirq/transformers/eject_phased_paulis.py (hunk 2 above)
Comment (tanujkhattar) [reply cirq-core/cirq/transformers/eject_phased_paulis.py]: done

File: cirq-core/cirq/transformers/eject_phased_paulis_test.py (hunk 3 above)
Comment (MichaelBroughton) [reply cirq-core/cirq/transformers/eject_phased_paulis_test.py:63]: SG

**Review by MichaelBroughton — APPROVED** [review]

**Pull request merged**